src/accumulator/proof.rs

```diff
@@ -137,7 +137,101 @@ impl Proof {
         }
         Ok(true)
     }
+    pub fn proof_after_deletion(
+        &self,
+        num_leafs: u64,
+    ) -> Result<(Vec<bitcoin_hashes::sha256::Hash>, Proof), String> {
+        let total_rows = util::tree_rows(num_leafs);
+        let mut targets = self.targets.to_owned();
+        targets.sort();
+
+        let proof_positions = util::get_proof_positions(&targets, num_leafs, total_rows);
+
+        let mut proof_hashes: Vec<_> = proof_positions
+            .to_owned()
+            .into_iter()
+            .zip(self.hashes.to_owned())
+            .collect();
+
+        let mut targets: Vec<_> = util::detwin(targets.to_owned(), total_rows)
+            .into_iter()
+            .rev()
+            .collect();
+
+        // Here is where our modified proof will be constructed
+        let mut target_hashes: Vec<(u64, bitcoin_hashes::sha256::Hash)> = vec![];
 
+        // For each of the targets, we'll try to find the sibling in the proof hashes
+        // and promote it as the parent. If it's not in the proof hashes, we'll move
+        // the descendatns of the existing targets and proofs of the sibling's parent
+        // up by one row.
+
+        while let Some(target) = targets.pop() {
+            if util::is_root_position(target, num_leafs, total_rows) {
+                target_hashes.push((target, bitcoin_hashes::sha256::Hash::default()));
+                continue;
+            }
+
+            let sib = target ^ 1;
+            // Is the node sibling on hash proofs?
+            if let Some(idx) = proof_hashes.iter().position(|(pos, _)| sib == *pos) {
+                let parent_pos = util::parent(sib as u64, total_rows);
+
+                target_hashes.push((parent_pos, proof_hashes[idx].1));
+
+                // Delete the sibling from proof_hashes as this sibling is a target now, not a proof.
+                proof_hashes.remove(idx);
+            } else {
+                // If the sibling is not in the proof hashes or the targets,
+                // the descendants of the sibling will be moving up.
+                //
+                // 14
+                // |---------------\
+                // 12              13
+                // |-------\       |-------\
+                // 08      09      10      11
+                // |---\   |---\   |---\   |---\
+                // 00  01          04  05  06  07
+
+                let to_update: Vec<_> = target_hashes
```

Comment:
We should also be updating the `proof_hashes` here or else the function won't work correctly. I don't think the current tests don't catch that.

Reply:
I changed this locally, but do you have test cases where it fails? It would be nice if we can test this before upstreaming.

Reply:
I took this from a failing fuzz test in the go library.

https://go.dev/play/p/ItlQAlQDxx3

There's two functions: one that updates the proof hashes and one that doesn't. The one that doesn't fails to verify

Reply:
The new commits fix it and adds this case in our tests. What do you think?

Reply:
Looks good :)